aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,356 @@
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_FRIEND_CLASS);
+#undef DEFINE_FRIEND_CLASS
+
+  using unboxed_iterator_type =
+      TORCH_ITENSORLIST_IMPL(Unboxed)::list_type::const_iterator;
+  using boxed_iterator_type =
+      TORCH_ITENSORLIST_IMPL(Boxed)::list_type::const_iterator;
+
+  union Payload {
+    boxed_iterator_type boxed_iterator;
+    unboxed_iterator_type unboxed_iterator;
+    void* _init_ptr;
+    Payload() : _init_ptr(nullptr) {}
+    ~Payload() = default;
+  };
+
+ public:
+  ITensorListIterator() : tag_(ITensorListTag::None) {}
+
+  ITensorListIterator(boxed_iterator_type boxed) : tag_(ITensorListTag::Boxed) {
+    payload_.boxed_iterator = boxed;
+  }
+
+  ITensorListIterator(unboxed_iterator_type unboxed)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed_iterator = unboxed;
+  }
+
+  detail::ITensorListConstRef operator*() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return ImplT::iterator_get(this_); });
+  }
+
+  ITensorListIterator& operator++() {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { ++this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator++(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { ++this_; });
+    return old;
+  }
+
+  ITensorListIterator& operator--() {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator--(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return old;
+  }
+
+  bool operator==(const ITensorListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ITENSORLIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const ITensorListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  Payload payload_;
+  ITensorListTag tag_;
+};
+
+/*
+ * [Note: ITensorList]
+ * Wrapper around boxed and unboxed API containers.
+ *
+ * Tagged union of both API containers:
+ *     - `TensorList`, a.k.a. `ArrayRef<Tensor>` (the unboxed API container)
+ *     - `List<Tensor>` (the boxed API container)
+ *
+ * This container wraps around these two, without incurring in extra overhead
+ * for converting from one to another.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ */
+class ITensorList {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...) friend class TORCH_ITENSORLIST_IMPL(TAG);
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_FRIEND_CLASS);
+#undef DEFINE_FRIEND_CLASS
+
+  using unboxed_type = TORCH_ITENSORLIST_IMPL(Unboxed)::list_type;
+  using boxed_type = TORCH_ITENSORLIST_IMPL(Boxed)::list_type;
+
+  union Payload {
+    const boxed_type* boxed;
+    unboxed_type unboxed;
+    Payload() : boxed(nullptr) {}
+    ~Payload() = default;
+  };
+
+ public:
+  using iterator = ITensorListIterator;
+  using const_iterator = ITensorListIterator;
+  using value_type = typename iterator::value_type;
+
+  ITensorList() : tag_(ITensorListTag::None) {}
```

Comment:
We should consider not allowing this to be default constructed; not allowing default construction can be fairly miserable but it seems warranted here, where ITensorList should ONLY ever occur in argument type position.

Reply:
That sounds like a fair point. But, what if the user wishes to pass an empty `ITensorList`? Should we make it a private constructor hidden behind a static function: `ITensorList::make_empty()`?

Reply:
Well, all the ArrayRef constructors are still valid, so if you do `{}` I'm expecting to hit the ArrayRef case. Or does this not work for some reason?

Reply:
I think it does if we keep that [templated constructor](https://github.com/pytorch/pytorch/pull/67964#discussion_r756261866) implicit (instead of explicit, as I did when adding support for `Tensor?[]`). Otherwise, I think it only works with `{{}}`, which I don't particularly like.

Reply:
{} ought to work, I'm guessing the templated constructor should be implicit or it's not going to be BC.